src/gedi_subset/maapx.py

```diff
@@ -2,109 +2,188 @@
+    """
+    granule = granule.get("Granule", {})
+    resources = granule.get("OnlineResources", {}).get("OnlineResource", [])
+    # OnlineResources.OnlineResource is a list only if there are multiple
+    # resources, so when there is only one resource, we must put it in a list.
+    # This is due to how the XML is parsed by the `xmltodict` library.
+    resources = resources if isinstance(resources, list) else [resources]
+
+    # We assume there is only one S3 credentials endpoint for a granule.
+    if s3_resource := next(filter(_is_s3_credentials_online_resource, resources), None):
+        return s3_resource["URL"]
+
+    return None
 
 
-@ttl_cache(ttl=55 * 60)
-def _earthdata_s3_credentials(maap: MAAP, endpoint: str) -> IOResultE["AWSCredentials"]:
-    """Returns short-term AWS credentials obtained from an S3 credentials endpoint."""
+@cached(cache=TTLCache(maxsize=1, ttl=55 * 60), key=lambda _, endpoint: endpoint)
+def _s3_credentials(maap: MAAP, endpoint: str) -> AWSCredentials | Never:
+    """Obtain short-term AWS credentials from an S3 credentials endpoint."""
 
     logger.debug(f"Obtaining S3 credentials from {endpoint}")
 
-    return impure_safe(maap.aws.earthdata_s3_credentials)(endpoint)
+    return maap.aws.earthdata_s3_credentials(endpoint)
 
 
 @cached(cache=FIFOCache(maxsize=1), key=lambda creds: creds["sessionToken"])
-def _setup_default_boto3_session(creds: "AWSCredentials") -> None:
-    """Sets up the default boto3 session using the specified credentials."""
+def _setup_default_boto3_session(creds: AWSCredentials) -> None:
+    """Set up the default boto3 session using the specified credentials."""
 
     logger.debug("Setting up default boto3 session with new credentials")
 
     boto3.setup_default_session(
         aws_access_key_id=creds["accessKeyId"],
         aws_secret_access_key=creds["secretAccessKey"],
         aws_session_token=creds["sessionToken"],
-        # TODO: make this configurable
         region_name="us-west-2",
+        botocore_session=botocore.session.Session(),
     )
 
 
-def download_granule(maap: MAAP, todir: str, granule: Granule) -> IOResultE[str]:
+def _fatal_request_error(e: Exception) -> bool:
+    _RETRYABLE_STATUS_CODES = {
+        408,  # Request Timeout
+        429,  # Too Many Requests
+        500,  # Internal Server Error
+        502,  # Bad Gateway
+        503,  # Service Unavailable
+        504,  # Gateway Timeout
+    }
+
+    if isinstance(e, requests.exceptions.RequestException):
+        status_code = e.response.status_code if e.response is not None else 0
+        return status_code not in _RETRYABLE_STATUS_CODES
+
+    return True
+
+
+def _make_retrier(max_tries=10) -> Callable[[Callable[P, T]], Callable[P, T]]:
+    return backoff.on_exception(
+        backoff.expo,
+        requests.exceptions.RequestException,
+        giveup=_fatal_request_error,
+        max_tries=max_tries,
+        jitter=lambda wait: random.uniform(0.8 * wait, 1.2 * wait),
```

Comment:
This `backoff` library is really cool, I wasn't aware of its existence :)

It's not super easy to understand what the successive wait times are going to be in theory, but let me attempt.

If I understand correctly, everything is based [on their `expo` function](https://github.com/litl/backoff/blob/d82b23c42d7a7e2402903e71e7a7f03014a00076/backoff/_wait_gen.py#L8). With a `factor=10` like we specify by default, it looks to me that our wait times non jittered will be : 

```
10*2**0 = 10
10*2**1 = 20
10*2**2 = 40
...
10*2**9 = 5120 
```

And our `jitter` value says we randomly pick a value around *0.8 and 1.2 times the given wait time, with equal chances.

The last values sound like a lot of time to me but, again, I am not familiar with the behavior of the code: how likely is it that this will fail 10 times (i.e how often does a granule download fail continuously during several hours). 

I was curious why we don't use the default `jitter` but it looks like it results in significant chances to have wait times close to zero which I guess we want to avoid.

Reply:
Yes, your understanding is spot on. Regarding your other comment about 10 retries with potentially long wait times, your observation there is correct as well. This is indeed a balancing act between robustness and speed. In this case, I chose to allow for more and longer wait times due to the specific nature of the transient errors we're trying to handle here.

The motivation for this is a flaky HTTP server that can sometimes require a bit of patience and persistence before we can get a successful download. This doesn't necessarily happen often, but it crops up often enough to become very frustrating to users when a single download failure causes the entire job to fail. See more in my reply to your other comment about how job failure is impacted.